featuregate/gates.go

```diff
@@ -19,12 +19,77 @@ import (
 	"sync"
 )
 
-// Gate represents an individual feature that may be enabled or disabled based
+// Stage represents the Gate's lifecycle and what is the expected state of it.
+type Stage int8
+
+const (
+	// Alpha is used when creating a new feature and the Gate must be explicitly enabled
+	// by the operator.
+	//
+	// The Gate will be disabled by default.
+	Alpha Stage = iota
+	// Beta is used when the feature flag is well tested and is enabled by default,
+	// but can be disabled by a Gate.
+	//
+	// The Gate will be enabled by default.
+	Beta
+	// Stable is used when feature is permanently enabled and can not be disabled by a Gate.
+	// This value is used to provide feedback to the user that the gate will be removed in the next version.
+	//
+	// The Gate will be enabled by default and will return an error if modified.
+	Stable
+)
+
+var stageNames = map[Stage]string{
+	Alpha:  "Alpha",
+	Beta:   "Beta",
+	Stable: "Stable",
+}
+
+func (s Stage) String() string {
+	if v, ok := stageNames[s]; ok {
+		return v
+	}
+	return "unknown"
+}
+
+// Gate is an immutable object that is owned by the `Registry`
+// to represents an individual feature that may be enabled or disabled based
 // on the lifecycle state of the feature and CLI flags specified by the user.
 type Gate struct {
-	ID          string
-	Description string
-	Enabled     bool
+	ID             string
+	Description    string
+	Enabled        bool
+	Stage          Stage
+	ReferenceURL   string
+	RemovalVersion string
```

Comment:
Let's start to make it immutable, by providing getters only and deprecate the public fields.

Reply:
Since the registry is returning a copy of the type, do you think it is valuable to do so?

I don't mind doing so just wanted to capture your thoughts here.

Reply:
If it is returning a copy, then changes in "IsEnabled" are not reflected correct? Is that important?

Reply:
> If it is returning a copy, then changes in "IsEnabled" are not reflected correct? 

No, so if make a call to `Registry.Apply` and update it after the gates have been read from `Registry.List` then you will have an out of date gate.

>  Is that important?

I would like to say no.
For example, if `zpages` lists each of the flags and then flags are then updated, reloading all of the gates to render that information resolves this.

However, it does raise the point of, should a gate have an exported method `IsEnabled` that could provide out of date information?